# 01. Import libraries

In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 02. Import data

In [2]:
#Import data
path = r'C:\Users\opetr\Documents\Instacart Basket Analysis'

In [4]:
#Import products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data',  'products.csv'), index_col = False)

In [9]:
#Import orderswrangled.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 03. Data consistency checks

## Mixed-type data

In [10]:
#Create a dataframe
df_test = pd.DataFrame()

In [11]:
#Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [12]:
#Check on test dataframe
df_test.head()

mix
0     a
1     b
2     1
3  True

In [13]:
#Check if df contains mixed-type columns
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [14]:
#Change data type to string
df_test['mix'] = df_test['mix'].astype('str')

## Missing values

In [15]:
#Finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [17]:
#Create subset to show missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [18]:
#Check on newly created subset
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [19]:
#Check initial number of rows in the df
df_prods.shape

(49693, 5)

In [20]:
#Create new clean df
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [21]:
#Check shape after change
df_prods_clean.shape

(49677, 5)

## Duplicates

In [22]:
#Finding duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [23]:
#Check new duplicate df
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [24]:
#Check initial number of rows in the df
df_prods_clean.shape

(49677, 5)

In [25]:
#Create new df that excludes duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [27]:
#Check shape after change
df_prods_clean_no_dups.shape

(49672, 5)

## Exporting changes

In [28]:
#Export new df
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# TASK 4.5

In [29]:
#2. Run the df.describe() function on your df_prods dataframe
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

#### The max price for a product is 99,999 which could be mistake since the mean is 9.99 and these are grocery store products that do not tend to be this expensive. I would communicate finding to client.

In [30]:
#3. Check for mixed-type data in your df_ords dataframe
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [ ]:
#4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type
#Answer: No mixed-type data found in this dataframe

In [31]:
#5. Run a check for missing values in your df_ords dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [34]:
df_ords.user_id.max()

206209

#### There are 206,209 missing values in the days_since_prior_order column. Since there are 206,209 individual User IDs in the user_id column, I hypothesize that that each User ID has an average of 1 entry missing in the days_since_prior_order column (one user_id might have 2 missing, whereas another might have none). It is also possible and more plausible that each User has a value missing due to no prior order from when they were a first time customer

In [37]:
#6. Address the missing values using an appropriate method
#Create a subset for missing values
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [38]:
#Call on new subset and inspect missing values as it relates to user_id
df_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

#### We can see the user_id goes in an uninterrupted sequence from 1 to 206,209, proving that each customer has one missing value in the column in question. The observations corresponding to each missing value could be useful in analyzing what a first time customer purchases. For this reason, we should not remove them. We should not impute a new value based on the mean or median due to the same reason - this information could be useful farther down the road. We could create a new variable that acts like a flag based on the missing value, but I think the best decision in this case would be to create a value of zero instead, showing that there were no prior purchases made. This would also make these observations easier to call on in the future, and it will be easy for anyone inspecting the data in the future to draw accurate conclusions.

In [41]:
#Update missing values to 0
df_ords['days_since_prior_order'].fillna(0, inplace=True)

In [44]:
#Check that each User ID has no missing values and one value of 0
df_ords.head(30)

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0            0   2539329        1             1                   2   
1            1   2398795        1             2                   3   
2            2    473747        1             3                   3   
3            3   2254736        1             4                   4   
4            4    431534        1             5                   4   
5            5   3367565        1             6                   2   
6            6    550135        1             7                   1   
7            7   3108588        1             8                   1   
8            8   2295261        1             9                   1   
9            9   2550362        1            10                   4   
10          10   1187899        1            11                   4   
11          11   2168274        2             1                   2   
12          12   1501582        2             2                   5   
13          13   1901567        2             3                   1   
14          14    738281        2             4                   2   
15          15   1673511        2             5                   3   
16          16   1199898        2             6                   2   
17          17   3194192        2             7                   2   
18          18    788338        2             8                   1   
19          19   1718559        2             9                   2   
20          20   1447487        2            10                   1   
21          21   1402090        2            11                   1   
22          22   3186735        2            12                   1   
23          23   3268552        2            13                   4   
24          24    839880        2            14                   3   
25          25   1492625        2            15                   1   
26          26   1374495        3             1                   1   
27          27    444309        3             2                   3   
28          28   3002854        3             3                   3   
29          29   2037211        3             4                   2   

    order_hour_of_day  days_since_prior_order  
0                   8                     0.0  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0  
11                 11                     0.0  
12                 10                    10.0  
13                 10                     3.0  
14                 10                     8.0  
15                 11                     8.0  
16                  9                    13.0  
17                 12                    14.0  
18                 15                    27.0  
19                  9                     8.0  
20                 11                     6.0  
21                 10                    30.0  
22                  9                    28.0  
23                 11                    30.0  
24                 10                    13.0  
25                 11                    30.0  
26                 14                     0.0  
27                 19                     9.0  
28                 16                    21.0  
29                 18                    20.0

In [46]:
#7. Run a check for duplicate values in your df_ords data
df_ords_dups = df_ords[df_ords.duplicated()]

In [48]:
#Check on duplicate subset
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

##### Address the duplicates using an appropriate method. Answer: There are no duplicate values in df_ords

In [50]:
#9a. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder
##df_prods exported before Task began
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_cleaned.csv'))